Accept a filename

vstinner opened the pull request:
* Rename patch() to download()
* Process patch content as Unicode: always assume UTF-8 encoding
* rename() uses a regular expression
* When a filename is used, rename() matchs the original name using a
  regular expression.
* On failure, store the patch as "ferrypick.patch" in the current
  diretory
* Remove unused imports

hroncok commented:
Why have you added executable bit and shebang as well?

hroncok commented on ferrypick.py:
```diff
@@ -1,15 +1,19 @@
 PR_RE = re.compile(r"^https://src\.fedoraproject\.org/\S+/([^/\s]+)/pull-request/\d+")
-REPLACE_SUFFIXES = "spec", "rpmlintrc"
+# filename without extension: 'python3.6'
+FILENAME_RE = r'[a-zA-Z0-9_.]+'
```
also can contain dashes and pluses https://docs.fedoraproject.org/en-US/packaging-guidelines/Naming/#_common_character_set_for_package_naming

hroncok commented on ferrypick.py:
```diff
@@ -1,15 +1,19 @@
+# This is what git does: "a" in "a/python37.spec"
+PREFIXES = ("a", "b")
+# Always assume UTF-8
+ENCODING = "utf-8"
```
Technically, the patch may have anything in it. But realistically, this will work in 99.9999 % cases. Is working on bytes so painful?

vstinner replied:
I changed the code to work on bytes again.

hroncok commented on ferrypick.py:
```diff
@@ -23,53 +27,87 @@ def parse_link(link):
-def rename(bytesline, original_name, current_name):
+def rename(content, original_name, current_name):
     """
     On a given bytes line, naïvely replace original package name with current package name.
```
This docstring is not accurate any more.

hroncok replied:
```suggestion
    In a given bytes patch-content, replace original package name with current package name.
    If original_name is None, it replaces a more general regular expression instead.
```

hroncok commented on ferrypick.py:
```diff
@@ -23,53 +27,87 @@ def parse_link(link):
+        name = re.escape(original_name)
+    else:
+        name = FILENAME_RE
+    original = (r"(?P<prefix>%s)/%s\.(?P<suffix>%s)"
+                % ('|'.join(PREFIXES),
+                   name,
+                   '|'.join(SUFFIXES)))
```
I can't read this at all.

vstinner replied:
It gives something like `a/python.spec` but in a more generic form to match also "b/" prefix or ".rpmlint" suffix :-)

At the end, it's closer to `(a|b)/python\.(spec|rpmlint)`.

Regex are fun!

hroncok replied:
It's actually the string modulo here that makes it hard to read for me. What about storing the prefixes regex-snippet and the suffixes regex-snippet in a constant and using an f-string here:

```python
rf"(?P<prefix>{PREFIXES_RE})/{name}\.(?P<suffix>{SUFFIXES_RE})"
```

Also, while converting this to regex this, could you wrap it in `\b`s? I haven't done that before only because I avoided regex.

hroncok commented on ferrypick.py:
```diff
@@ -23,53 +27,87 @@ def parse_link(link):
-            tmp_file.flush()
-        yield tmp_file.name
+    with urllib.request.urlopen(link) as response:
+        conte = response.read()
```
```suggestion
        content = response.read()
```

hroncok commented on ferrypick.py:
```diff
@@ -23,53 +27,87 @@ def parse_link(link):
+    print(f"$ {cmd}")
+    exitcode = execute(cmd)
+    if exitcode:
+        print()
+        print(f"{cmd} failed with exit code {exitcode}")
+        print(f"Patch stored as: {filename}")
```
Print this to stderr?

vstinner commented:
> Why have you added executable bit and shebang as well?

I like to run ~/dev/ferrypick/ferrypick.py directly without having to install it.

hroncok commented:
> I like to run ~/dev/ferrypick/ferrypick.py directly without having to install it.

You can run it with `python ~/dev/ferrypick/ferrypick.py`.

Disclaimer: I don't like storing shebangs in sources. If you use `/usr/bin/python3` and run the script directly with a venv activated, it breaks user assumptions. If at all, I'd prefer the `/usr/bin/env python3` shebang, but I'd rather not have it at all. Shebangs are tricky.

vstinner commented:
I addresses all of your mark. Please review the updated PR.

hroncok commented on ferrypick.py:
```diff
@@ -1,15 +1,17 @@
 COMMIT_RE = re.compile(r"^https://src\.fedoraproject\.org/\S+/([^/\s]+)/c/([0-9a-f]+)")
 PR_RE = re.compile(r"^https://src\.fedoraproject\.org/\S+/([^/\s]+)/pull-request/\d+")
-REPLACE_SUFFIXES = "spec", "rpmlintrc"
+# filename without extension: 'python3.6'
+# https://docs.fedoraproject.org/en-US/packaging-guidelines/Naming/#_common_character_set_for_package_naming
+FILENAME_RE = r'[a-zA-Z0-9_.+-]+'
```
What if this is called PKGNAME_RE? That way, it actually is, what it is.

vstinner commented:
I changed how the final regex is built.

> Also, while converting this to regex this, could you wrap it in \bs? I haven't done that before only because I avoided regex.

I understand what wrapping in `\b` means. Do you mean to change a regex? Which one?

hroncok commented on ferrypick.py:
```diff
@@ -1,15 +1,18 @@
 COMMIT_RE = re.compile(r"^https://src\.fedoraproject\.org/\S+/([^/\s]+)/c/([0-9a-f]+)")
 PR_RE = re.compile(r"^https://src\.fedoraproject\.org/\S+/([^/\s]+)/pull-request/\d+")
-REPLACE_SUFFIXES = "spec", "rpmlintrc"
+# filename without extension: 'python3.6'
```
```suggestion
```

hroncok commented on ferrypick.py:
```diff
@@ -1,15 +1,18 @@
+# filename without extension: 'python3.6'
+# https://docs.fedoraproject.org/en-US/packaging-guidelines/Naming/#_common_character_set_for_package_naming
+PKGNAME_RE = r'[a-zA-Z0-9_.+-]+'
+SUFFIXES_RE = r"\.(spec|rpmlintrc)"
```
```suggestion
# Files named pkgname.spec and pkgname.rpmlintrc need to be renamed in patches
SUFFIXES_RE = r"\.(spec|rpmlintrc)"
```

hroncok commented on ferrypick.py:
```diff
@@ -1,15 +1,18 @@
+SUFFIXES_RE = r"\.(spec|rpmlintrc)"
+# This is what git does: "a" in "a/python37.spec"
+PREFIXES_RE = r'(a|b)'
+RENAME_RE = f"(?P<prefix>{PREFIXES_RE})/%s(?P<suffix>{SUFFIXES_RE})"
```
```suggestion
RENAME_RE_TEMPLATE = f"(?P<prefix>{PREFIXES_RE})/{{}}(?P<suffix>{SUFFIXES_RE})"
```

hroncok commented on ferrypick.py:
```diff
@@ -23,53 +26,85 @@ def parse_link(link):
+        name_regex = re.escape(original_name)
+    else:
+        name_regex = PKGNAME_RE
+    regex = RENAME_RE % name_regex
```
```suggestion
    regex = RENAME_RE_TEMPLATE.format(name_regex)
```

hroncok commented on ferrypick.py:
```diff
@@ -23,53 +26,85 @@ def parse_link(link):
-    with patch(patch_link, original_name, current_name) as p:
-        print(f"$ git am --reject {p}")
-        execute(f"git am --reject {p}")
+        current_name = stdout("git rev-parse --show-toplevel")
+        current_name = os.path.basename(current_name)
```
```suggestion
        git_toplevel = stdout("git rev-parse --show-toplevel")
        current_name = os.path.basename(git_toplevel)
```

hroncok commented on ferrypick.py:
```diff
@@ -23,53 +26,85 @@ def parse_link(link):
+        current_name = stdout("git rev-parse --show-toplevel")
+        current_name = os.path.basename(current_name)
+
+    if os.path.exists(link):
+        with open(link, 'rb') as fp:
+            content = fp.read()
+        original_name = None
+    else:
+        original_name, patch_link = parse_link(link)
+        content = download(patch_link)
```
This IMHO deserves it's own function now.

vstinner commented:
@hroncok: I addressed suggestions of my latest review.

hroncok commented:
Do the test still pass? Would you mind adding some new? (There is no tests doc yet, sorry.)

vstinner commented:
> Do the test still pass? Would you mind adding some new? (There is no tests doc yet, sorry.)

Oh. I didn't notice that there were tests. I fixed tests. I converted the last one to a more realistic "functional tests" (but skip download... since Pagure is currently down :-) ... and git am).

vstinner commented:
I created https://github.com/fedora-python/ferrypick/pull/2 to run the tests on Travis CI.

torsava commented:
Omg, I was about to implement this right now. Thanks @vstinner and @hroncok :)
